만국박람회 [STEP 2] cheshire.kwon, claire.pp

hano-ios opened the pull request:
안녕하세요 @wonhee009 
만국박람회 STEP 2 PR입니다

저희가 STEP2를 하며 구현한 것은 다음과 같습니다
- 데이터 타입 변경
- 화면 UI 구성(메인, 출품작 리스트, 출품작 설명 화면)
- 네비게이션 컨트롤러를 통한 화면 이동
- parsing한 JSON data를 이용한 데이터 출력

## 데이터 타입 변경
STEP 1에서 코멘트 주신 내용(Codable)을 보고 토론해본 결과
Codable(Encodable + Decodable)에서 저희는 Decodable만 필요했기 때문에
EntryData와 ExpositionData를 Decodable 프로토콜을 채택하도록 수정했습니다

## 화면 UI 구성
저희는 코드로 화면을 구성하였습니다.
화면에 대한 각각의 뷰컨트롤러 내부에 스택뷰를 이용하여 뷰를 만들고,
스크롤뷰가 필요한 경우, 해당 뷰를 감싸는 형식으로 구현했습니다.

## 네비게이션 컨트롤러를 통한 화면 이동
저희는 전체 프로젝트를 네비게이션 컨트롤러로 활용하고자 했습니다.
그래서 메인 -> 한국의 출품작 -> <출품작명>
의 타이틀을 가지는 뷰 세개가 있다고 생각하고 프로젝트를 진행했습니다.
따라서 네비게이션 컨트롤러를 사용하기 위해서 <code>SceneDelegate</code>에
```Swift
        let rootViewController = MainViewController()
        let navigationController = UINavigationController(rootViewController: rootViewController)
        window?.rootViewController = navigationController
        window?.makeKeyAndVisible()
```
와 같이 구현하였고,
```Swift
        self.navigationController?.pushViewController(secondView, animated: true)
```
의 코드를 통해 화면 이동을 구현했습니다.

## Parsing한 JSON 데이터 출력
테이블뷰의 각 행에 맞는 <code>EntryData</code>의 정보를 출력하기 위해서
각 Cell에 <code>makeEntryCell(data: EntryData)</code> 함수를 만들고,
테이블뷰에서 셀에 해당 함수를 통해 JSON 데이터를 보내고, 셀은 내부의 imageView와 label의 값을 바꾸는 형식으로 구현했습니다.

## 질문
- MainViewController에서 scrollView의 constraint를 view.layoutMarginsGuide에 맞게 했는데, 앱 실행 화면에서는 safeArea 안쪽으로 맞게 화면이 나와서 어떻게 해결할 수 있는지 궁금합니다
- Cell의 초기화 과정에서 처음에는 <code> awakeFromNib </code>에서 각 내부 요소들의 constraint를 잡아주는 등의 함수를 실행했는데, 실제로 화면에 출력이 되지는 않았습니다. 그래서 <code> init </code>으로 모두 옮겨서 실행했더니 정상적으로 화면이 나왔습니다. 함수 설명에도  <code> awakeFromNib </code>는 초기화 후에 호출된다고 나왔는데, 왜  <code> awakeFromNib </code>에서는 제대로 실행되지 않았는지 궁금합니다.

Commit:
feat: [UI] 메인 화면 스크롤뷰 추가

메인 화면으로 띄울 스크롤 뷰 추가

Commit:
feat: 파리 만국박람회 화면 추가

파리 만국박람회 화면 - 정보 출력, 스크롤뷰, 버튼 추가

Commit:
feat: [UI] View 분리

MainViewController -> MainView 분리

Commit:
feat: 스크롤 뷰 추가...(미완성..)

Commit:
feat: 메인화면 구현, 출품작 이동

메인 화면 스크롤뷰 구현
출품작 리스트 화면으로 이동하여 테이블 뷰 띄워줌

Commit:
feat: [추가] 한국의 출품작 화면 구성

커스텀셀 만들어서 테이블 뷰 구성

Commit:
feat: 출품작 설명 화면 추가, 코드 정리, Data 타입 수정

출품작 화면 구현
의미없는 주석과 중복되는 코드 정리
Data type Codable -> Decodable로 수정

Commit:
chore: [수정] 코드 수정

?? -> if-let으로 수정

wonhee009 commented on Expo1900/Expo1900/Data/Decoder.swift:
```diff
@@ -0,0 +1,17 @@
+import UIKit
+
+class Decoder {
+    static func decodeEntry<T: Decodable> (type: T.Type, from: String) -> T? {
```
해당 메서드로 `expositionData`도 디코딩하고 있는데요.
현재 이름으로는 entry에 대한 것만 디코딩 가능하다고 알려주고 있는거 같아요.
제네릭을 사용해 원하는 데이터를 디코딩할 수 있게 공통화한 만큼 이름에서도 드러나면 더 좋을거 같습니다!😀

cheshire-kwon replied:
decodeJSONData로 수정했습니다!

wonhee009 commented on Expo1900/Expo1900/Data/Decoder.swift:
```diff
@@ -0,0 +1,17 @@
+    static func decodeEntry<T: Decodable> (type: T.Type, from: String) -> T? {
+        guard let items = NSDataAsset.init(name: from) else {
+            return nil
+        }
```
디코딩하려는 파일의 이름을 잘못 적으면 해당 가드문에서 nil을 반환할거 같은데요.
개발자가 디코딩이 필요한 코드를 작성하다 잘못 입력한 경우 혹은 이후 스텝에서 디코딩 파일 이름이 변경된 경우 등 여러 상황에서 해당 에러가 발생할 수 있을거 같습니다.
nil 보다는 error를 발생하는 편이 이후 에러 트랙킹할때 좀 더 유리할거 같은데요. 어떻게 생각하실까요?🤔

cheshire-kwon replied:
지적 감사합니다. 리턴이 아니라 에러로 변경하였습니다.!

wonhee009 commented on Expo1900/Expo1900/Data/Decoder.swift:
```diff
@@ -0,0 +1,17 @@
+import Foundation
+import UIKit
+
+class Decoder {
```
클래스로 선언하신 이유가 궁금합니다!

hano-ios replied:
상속도 하지 않고, 내부 변수를 바꿀 가능성이 없다고 생각해서 struct로 수정했습니다!

wonhee009 commented on Expo1900/Expo1900/Data/EntryData.swift:
```diff
@@ -1,6 +1,6 @@
 import Foundation
 
-struct EntryData: Codable{
+struct EntryData: Decodable{
```
GOOD!👍

wonhee009 commented on Expo1900/Expo1900/View/EntryDetailView.swift:
```diff
@@ -0,0 +1,40 @@
+        label.font = UIFont.preferredFont(forTextStyle: .body)
+        label.numberOfLines = 0
+        return label
+    }()
```
이후에 수정 가능성이 있는 UI요소들인가요?
아니라면 `let`으로 선언은 어떨까요?

hano-ios replied:
let으로 수정했습니다! 감사합니다

wonhee009 commented on Expo1900/Expo1900/View/EntryDetailView.swift:
```diff
@@ -0,0 +1,40 @@
+    func makeDetailView(data: EntryData) {
+        entryImage.image = UIImage(named: data.image)
+        entryDescription.text = data.detailDescription
+    }
```
detailView를 설정하는데 필요한 데이터들을 외부에서 설정 가능하도록 하셨군요👍
해당 화면의 정보들은 한 번 생성되면 바뀔 가능성이 없어보이는데요.
init 시점에 데이터도 같이 받는건 어떨까요?
```swift
init(entry: EntryData) {
     super.init(frame: .zero)
     settingStackView()
     makeDetailView(entry)
}
```
이런식으로요.

wonhee009 replied:
또한 `make-`라는 네이밍이 뭔가 view를 생성해서 반환할거 같은데요.
UI요소들을 설정이나 지정해주는 역할을 하고 있으니 해당 역할을 나타낼 수 있는 이름은 어떨까요?😀

hano-ios replied:
EntryDetailViewController와 EntryDetailView의 init을 이용해서 초기화하도록 변경했습니다. 또한, setDetailView(data: EntryData)로 수정하여 네이밍을 조금 더 명확하게 바꿔보았습니다! 감사합니당

wonhee009 commented on Expo1900/Expo1900/View/EntryDetailView.swift:
```diff
@@ -0,0 +1,40 @@
+import UIKit
+
+class EntryDetailView: UIStackView {
```
`UIStackView`를 사용하셨네요👍
StackView를 사용하신 이유가 따로 있을까요?

cheshire-kwon replied:
스택뷰를 사용하면 contraint와 높이 등을 따로 설정해주지 않아도 되어서 편리하다고 판단했습니다.

wonhee009 commented on Expo1900/Expo1900/ViewController/EntryDetailViewController.swift:
```diff
@@ -0,0 +1,25 @@
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        view.backgroundColor = .white
```
backgroundColor를 white로 지정해주면 dark 모드에서는 어떻게 나오나요?🤔

hano-ios replied:
.systembackground로 수정했습니다. 다크모드는 생각하지 못했었는데 감사합니다!

wonhee009 commented on Expo1900/Expo1900/ViewController/EntryDetailViewController.swift:
```diff
@@ -0,0 +1,25 @@
+    func pass(data: EntryData) {
+        navigationItem.title = data.name
+        detailView.makeDetailView(data: data)
+    }
```
출품작 리스트에서 출품작을 클릭한 후 detail 뷰에 필요한 정보들을 전달해주기 위한 메서드로 이해했는데요.
pass보다는 `configure(entry: EntryData)`와 같은 네이밍이 좀 더 명확하지 않을까 생각하는데요.
어떻게 생각하실까요?🤔

hano-ios replied:
네이밍 추천 감사합니다! 앞부분 구조 수정하면서 해당 메서드는 삭제되었지만, 좋은 지적 감사합니다.

wonhee009 commented on Expo1900/Expo1900/ViewController/EntryDetailViewController.swift:
```diff
@@ -0,0 +1,25 @@
+        arrangeConstraint(view: scrollView, guide: view.layoutMarginsGuide)
+        arrangeConstraint(view: detailView, guide: scrollView.contentLayoutGuide)
+        detailView.widthAnchor.constraint(equalTo: scrollView.frameLayoutGuide.widthAnchor).isActive = true
+    }
```
해당 메서드도 EntryDetailVC에서만 사용될거 같은데요. `private` 선언해주는건 어떨까요?
또한 `create-`라는 네이밍이 UI를 만들어서 반환해야할거 같은 느낌인데요.
`setup`이나 `configure`같은 prefix를 많이 붙이는거 같습니다:)

hano-ios replied:
private setupUI()로 수정했습니다. 좋은 Prefix 추천해주셔서 감사합니다!!

wonhee009 commented on Expo1900/Expo1900/ViewController/EntryDetailViewController.swift:
```diff
@@ -0,0 +1,25 @@
+        view.addSubview(scrollView)
+        scrollView.addSubview(detailView)
+        arrangeConstraint(view: scrollView, guide: view.layoutMarginsGuide)
+        arrangeConstraint(view: detailView, guide: scrollView.contentLayoutGuide)
```
![image](https://user-images.githubusercontent.com/30456206/149127509-467960b9-ba46-466e-9e9f-c5c6a1337b6b.png)

스크롤바가 화면과 떨어져서 보이는 부분이 어색한거 같습니다.
`layoutMarginsGuide`에 맞춰서 constraints가 설정되어 그런데요.
scrollView는 superView 혹은 safeArea에 딱 맞추고 scrollView의 content 영역에 inset을 주는 방향은 어떨까요?

hano-ios replied:
scrollView를 safeArea에 맞추어 constraint를 설정했습니다.
그런데 scrollview.contentInset을 바꾸려 다음과 같은 코드를 이용했습니다.
<code> scrollView.contentInset = UIEdgeInsets(top: 0, left: 20, bottom: 0, right: 20) </code>
그런데 실행시 전체적인 컨텐츠가 확장되는 것 같아서 뷰가 짤리는 모습이 나왔습니다.
그래서 대신 스크롤뷰 내부의 MainView의 width를 작게 조정하고, 뷰의 centerX를 맞추어 구현했습니다.
이부분 어떤 부분이 잘못된것인지 알려주시면 감사하겠습니다!

wonhee009 replied:
음... 이 부분은 저도 다시 한 번 봐봐야할거 같습니다👀
오토레이아웃 관련 부분은 다음 스텝이니 거기서 좀 더 얘기해봐도 좋을거 같습니다!👍

wonhee009 commented on Expo1900/Expo1900/ViewController/EntryListViewController.swift:
```diff
@@ -0,0 +1,42 @@
+    }
+}
+
+extension EntryListViewController: UITableViewDelegate {
+    
+}
```
delegate 안에 구현해 놓은 메서드가 없는데요. 그럼 해당 코드를 지우면 어떻게 될까요?👀
[UITableViewDelegate](https://developer.apple.com/documentation/uikit/uitableviewdelegate)와 [UITableViewDataSource](https://developer.apple.com/documentation/uikit/uitableviewdatasource)에 어떤 메서드들이 있는지 확인해보고 각자의 위치가 맞는지 한 번 확인해보시면 좋을거 같습니다!

hano-ios replied:
UITableViewDelegate와 UITableViewDataSource의 함수를 확인했습니다!
DataSource 프로토콜 채택시 필수구현해야하는 두가지 메서드는 그대로 DataSource에 두었고,
func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) 함수는 Delegate로 이동했습니다
감사합니다!!

wonhee009 commented on Expo1900/Expo1900/ViewController/EntryListViewController.swift:
```diff
@@ -0,0 +1,42 @@
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        view.backgroundColor = .white
```
여기도 동일하게 dark 모드일때에는 어떻게 되나요?👀

hano-ios replied:
이부분도 위의 코멘트랑 동일하게 systembackground로 수정했습니다!!

wonhee009 commented on Expo1900/Expo1900/ViewController/EntryListViewController.swift:
```diff
@@ -0,0 +1,42 @@
+        super.viewDidLoad()
+        view.backgroundColor = .white
+        navigationItem.title = "한국의 출품작"
+        navigationController?.setNavigationBarHidden(false, animated: false)
```
메인 화면에서는 NavigationBar가 hidden 처리 했기때문에 해당 화면에서는 false로 만들어주고 있는거로 이해했습니다.
"메인->리스트->메인"으로 화면을 이동한다면 NavigationBar는 어떻게 될까요?👀

hano-ios replied:
정말 생각 못했던 부분이었는데 감사합니다!
NavigationBar를 유일하게 사용하지 않는 MainViewController의
viewWillAppear에서 NavigationBarHidden을 true로 하였고,
viewWillDisappear에서 NavigationBarHidden을 false로 하여
메인뷰에서만 네비게이션바가 없어지도록 했습니다!

wonhee009 commented on Expo1900/Expo1900/ViewController/EntryListViewController.swift:
```diff
@@ -0,0 +1,42 @@
+        view.addSubview(tableView)
+        arrangeConstraint(view: tableView, guide: view.layoutMarginsGuide)
+        tableView.delegate = self
+        tableView.dataSource = self
```
tableView를 설정하는 코드들은 별도의 메서드로 분리하는건 어떨까요?👀

hano-ios replied:
func setUpTableView() 함수에 별도로 정의했습니다!

wonhee009 commented on Expo1900/Expo1900/ViewController/EntryListViewController.swift:
```diff
@@ -0,0 +1,42 @@
+        let cell = EntryTableViewCell()
+        cell.makeEntryCell(data: entryItems[indexPath.row])
+        return cell
+    }
```
tableView의 모든 row에 대해서 새로 `EntryTableViewCell`을 생성해서 사용하고 있네요.👀
현재 화면 상에서는 스펙대로 잘 나오고 있습니다.
그렇다면 현재 코드에서 StackView를 사용하는 것과 TableView를 사용하는 것의 차이가 무엇인가요?
[UITableView](https://developer.apple.com/documentation/uikit/uitableview)에서 **Recycling Table View Cells**와 관련된 부분들 살펴보시면 좋을거 같습니다!
살펴 보시고 같이 얘기해봐요🙏

hano-ios replied:
<code> let cell = tableView.dequeueReusableCell(withIdentifier: EntryTableViewCell.identifier) as? EntryTableViewCell ?? EntryTableViewCell() </code>
의 구문을 이용해서 재사용했습니다! 감사합니다!!

wonhee009 commented on Expo1900/Expo1900/ViewController/EntryListViewController.swift:
```diff
@@ -0,0 +1,42 @@
+    }
+    
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        let thirdView = EntryDetailViewController()
```
`thirdView`보다는 의미 있는 네이밍은 어떨까요?😀

hano-ios replied:
secondView -> entryListView
thirdView -> entryDetailView
로 수정했습니다!

wonhee009 commented on Expo1900/Expo1900/ViewController/MainViewController.swift:
```diff
@@ -0,0 +1,38 @@
+        createUI()
+    }
+    
+    func createUI() {
+        view.addSubview(scrollView)
+        scrollView.addSubview(mainView)
+        arrangeConstraint(view: scrollView, guide: view.layoutMarginsGuide)
+        arrangeConstraint(view: mainView, guide: scrollView.contentLayoutGuide)
+        mainView.widthAnchor.constraint(equalTo: scrollView.frameLayoutGuide.widthAnchor).isActive = true
+    }
```
여기도 동일하게 `create`말고 다른 네이밍은 어떨까요?👀

hano-ios replied:
동일하게 setupUI()로 수정했습니다!!

wonhee009 commented on Expo1900/Expo1900/ViewController/MainViewController.swift:
```diff
@@ -0,0 +1,38 @@
+    func toKoreanEntries() {
+        let secondView = EntryListViewController()
+        self.navigationController?.pushViewController(secondView, animated: true)
+    }
```
`moveKoreanEntries()`는 어떨까요?

hano-ios replied:
네이밍 추천 감사합니다!! moveKoreanEntries()로 수정했습니다!!

wonhee009 commented on Expo1900/Expo1900/View/MainView.swift:
```diff
@@ -0,0 +1,95 @@
+    }()
+    
+    private var expositionImageView: UIImageView = {
+        let imageView: UIImageView = UIImageView()
+        imageView.image = UIImage(named: "poster")
+        imageView.contentMode = .scaleAspectFit
+        return imageView
+    }()
+    
+    private var expositionInfomation: UILabel = {
+        let label = UILabel()
+        label.textAlignment = .center
+        label.font = UIFont.preferredFont(forTextStyle: .title3)
+        label.numberOfLines = 0
+        return label
+    }()
+    
+    private var expositionDescription: UILabel = {
+        let label = UILabel()
+        label.font = UIFont.preferredFont(forTextStyle: .body)
+        label.numberOfLines = 0
+        return label
+    }()
+    
+    private lazy var koreanButtonView: UIStackView = {
+        let stackView = UIStackView(arrangedSubviews: [makeFlag(), koreanEntries, makeFlag()])
+        stackView.axis = .horizontal
+        stackView.distribution = .equalCentering
+        stackView.spacing = 20
+        stackView.alignment = .center
+        return stackView
+    }()
+    
+    private var koreanEntries: UIButton = {
+        let button = UIButton()
+        button.setTitle("한국의 출품작 보러가기", for: .normal)
+        button.setTitleColor(.systemBlue, for: .normal)
+        button.addTarget(self, action: #selector(MainViewController.toKoreanEntries), for: .touchUpInside)
+        return button
+    }()
```
중간에 수정될 UI 요소들이 아니라면 `let`으로 선언하는건 어떨까요?😀

hano-ios replied:
수정될 가능성이 없기 때문에 let으로 변경했습니다!

wonhee009 commented on Expo1900/Expo1900/View/MainView.swift:
```diff
@@ -0,0 +1,95 @@
+        return label
+    }()
+    
+    private lazy var koreanButtonView: UIStackView = {
+        let stackView = UIStackView(arrangedSubviews: [makeFlag(), koreanEntries, makeFlag()])
+        stackView.axis = .horizontal
+        stackView.distribution = .equalCentering
+        stackView.spacing = 20
+        stackView.alignment = .center
+        return stackView
+    }()
```
`makeFlag()` 때문에 `lazy`가 붙은거로 이해했는데요.
모든 UI 요소를 미리 만들 필요는 없습니다.
```swift
private let koreanButtonStackView: UIStackView = { ... }()

private func configureKoreanButtonStackView() {
     koreanButtonStackView.addSubview(makeFlag())
}
```

hano-ios replied:
lazy를 사용하면 미리 만들어놓는 것이 아니라, 필요한 시점에 생성한다고 알고 있습니다.
그래서 lazy를 사용하는 것이 일반적인 var를 사용하는 것보다 효율적이라고 생각했었는데 lazy 사용을 지양하는 것이 좋은 것인가요..?
또, 라자냐가 예시로 주신 코드를 이용해보려했는데, 
```Swift
    private func addFlag() {
        koreanButtonView.addSubview(makeFlag())
        koreanButtonView.addSubview(koreanEntries)
        koreanButtonView.addSubview(makeFlag())
    }
```
잘 동작하지 않아서 어떤 점이 잘못됐는지도 궁금합니다.
(init에서 addFlag를 호출했습니다.)

정리하자면,
- lazy대신 let을 사용하고 다른 함수를 이용하여 추가해야 하는 이유
- 위의 코드가 제대로 동작하지 않는 이유

가 궁금합니다!!

wonhee009 replied:
lazy를 지양한다기 보다는 적절한 곳에 잘 쓰였냐의 관점에서 질문 드렸었습니다.
lazy가 필요한 시점에 생성하는건 맞는데 화면을 구성하는 UI로써 초기화 시점에 생성되어야 한다고 생각했습니다.
물론, UI 요소 중에서도 lazy가 필요한 경우가 있는데요.
예를 들어, 데이터 값을 보고 UI의 설정을 바꿔줘야하는 경우가 있을거 같네요.

적어주신 코드가 제대로 동작하지 않은 이유에 대해서는 `.addSubview`를 호출했기 때문인데요.
```swift
private let koreanButtonView: UIStackView = {
        let stackView = UIStackView()
        stackView.translatesAutoresizingMaskIntoConstraints = false
        stackView.distribution = .equalCentering
        stackView.spacing = 20
        stackView.alignment = .center
        return stackView
    }()

koreanButtonView.addArrangedSubview(makeFlag())
koreanButtonView.addArrangedSubview(koreanEntries)
koreanButtonView.addArrangedSubview(makeFlag())
```
제가 제안드렸던 코드의 모습은 위와 같습니다.
UIStackView에서 `addArrangedSubView` 메서드에 대해서 한 번 보시고 `addSubView`와의 차이에 대해서도 한 번 보시면 좋을거 같습니다:)

wonhee009 commented on Expo1900/Expo1900/View/EntryTableViewCell.swift:
```diff
@@ -0,0 +1,86 @@
+
+class EntryTableViewCell: UITableViewCell {
+    
+    static let identifier = "EntryTableViewCell"
```
`identifier`를 선언해두고 사용하고 있지는 않네요🤔
왜 선언해 두셨는지 궁금하고, 어디에 사용할 수 있을까요?

hano-ios replied:
셀 재사용을 위해 넣었다가 사용하지 않았었습니다
identifier이용해서 셀 재사용 구현했습니다!!

wonhee009 commented on Expo1900/Expo1900/View/EntryTableViewCell.swift:
```diff
@@ -0,0 +1,86 @@
+        label.text = ">"
+        label.textColor = .gray
+        return label
+    }()
```
[이런 것도](https://developer.apple.com/documentation/uikit/uitableviewcell/1623228-accessorytype) 활용해 볼 수 있을거 같습니다:)

hano-ios replied:
꿀팁 감사합니다..ㅎㅎㅎ 수정했습니다!!

wonhee009 commented on Expo1900/Expo1900/View/EntryTableViewCell.swift:
```diff
@@ -0,0 +1,86 @@
+    }()
+    
+    lazy var entryInfomationStack: UIStackView = {
+        let stackView = UIStackView(arrangedSubviews: [entryImage,textStack])
```
```suggestion
        let stackView = UIStackView(arrangedSubviews: [entryImage, textStack])
```

hano-ios replied:
꼼꼼한 코멘트 감사합니다.. 수정했습니다!!

wonhee009 commented on Expo1900/Expo1900/View/EntryTableViewCell.swift:
```diff
@@ -0,0 +1,86 @@
+    }()
+    
+    lazy var entryStack: UIStackView = {
+        let stackView = UIStackView(arrangedSubviews: [entryInfomationStack,arrow])
```
```suggestion
        let stackView = UIStackView(arrangedSubviews: [entryInfomationStack, arrow])
```

hano-ios replied:
수정했습니다!!

wonhee009 commented on Expo1900/Expo1900/View/EntryTableViewCell.swift:
```diff
@@ -0,0 +1,86 @@
+        return label
+    }()
+    
+    lazy var textStack: UIStackView = {
+        let stackView = UIStackView(arrangedSubviews: [entryName, entryShortDescription])
+        stackView.translatesAutoresizingMaskIntoConstraints = false
+        stackView.axis = .vertical
+        stackView.distribution = .equalCentering
+        return stackView
+    }()
+    
+    lazy var entryInfomationStack: UIStackView = {
+        let stackView = UIStackView(arrangedSubviews: [entryImage,textStack])
+        stackView.translatesAutoresizingMaskIntoConstraints = false
+        stackView.axis = .horizontal
+        stackView.distribution = .fill
+        stackView.spacing = 10
+        stackView.alignment = .center
+        
+        return stackView
+    }()
+    
+    lazy var entryStack: UIStackView = {
+        let stackView = UIStackView(arrangedSubviews: [entryInfomationStack,arrow])
+        stackView.translatesAutoresizingMaskIntoConstraints = false
+        stackView.axis = .horizontal
+        stackView.distribution = .equalSpacing
+        stackView.alignment = .center
```
`lazy`로 선언한 이유가 있을까요?👀

hano-ios replied:
stackView에 다른 요소들을 넣기 위해서는 lazy var를 사용해야한다고 알고 있었습니다
다른 요소들이 생성되는 시점이 아닌 실제 사용되는 시점에 생성하기 위해서 lazy를 이용했습니다

wonhee009 commented on Expo1900/Expo1900/View/EntryTableViewCell.swift:
```diff
@@ -0,0 +1,86 @@
+        return stackView
+    }()
+    
+    func makeEntryCell(data: EntryData) {
```
이 부분도 `make`가 아닌 다른 네이밍은 어떨까요?:)

hano-ios replied:
setEntryCell로 수정했습니다!!

wonhee009 commented on Expo1900/Expo1900/View/EntryTableViewCell.swift:
```diff
@@ -0,0 +1,86 @@
+        entryShortDescription.text = data.shortDescription
+    }
+    
+    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
+        super.init(style: style, reuseIdentifier: reuseIdentifier)
+        contentView.addSubview(entryStack)
+        
+        entryImage.widthAnchor.constraint(equalTo: contentView.widthAnchor, multiplier: 0.2).isActive = true
+        entryImage.heightAnchor.constraint(equalTo: entryImage.widthAnchor).isActive = true
+        
+        entryStack.topAnchor.constraint(equalTo: contentView.layoutMarginsGuide.topAnchor).isActive = true
+        entryStack.bottomAnchor.constraint(equalTo: contentView.layoutMarginsGuide.bottomAnchor).isActive = true
+        entryStack.leadingAnchor.constraint(equalTo: contentView.layoutMarginsGuide.leadingAnchor).isActive = true
+        entryStack.trailingAnchor.constraint(equalTo: contentView.layoutMarginsGuide.trailingAnchor).isActive = true
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
```
메서드 최상단으로 이동은 어떨까요?
이런 메서드의 위치도 하나의 흐름을 개략적으로 보여줄 수 있다고 생각합니다!

hano-ios replied:
참조해서 메서드 위치 수정했습니다!!

wonhee009 commented on Expo1900/Expo1900/View/EntryTableViewCell.swift:
```diff
@@ -0,0 +1,86 @@
+    
+    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
+        super.init(style: style, reuseIdentifier: reuseIdentifier)
+        contentView.addSubview(entryStack)
+        
+        entryImage.widthAnchor.constraint(equalTo: contentView.widthAnchor, multiplier: 0.2).isActive = true
+        entryImage.heightAnchor.constraint(equalTo: entryImage.widthAnchor).isActive = true
+        
+        entryStack.topAnchor.constraint(equalTo: contentView.layoutMarginsGuide.topAnchor).isActive = true
+        entryStack.bottomAnchor.constraint(equalTo: contentView.layoutMarginsGuide.bottomAnchor).isActive = true
+        entryStack.leadingAnchor.constraint(equalTo: contentView.layoutMarginsGuide.leadingAnchor).isActive = true
+        entryStack.trailingAnchor.constraint(equalTo: contentView.layoutMarginsGuide.trailingAnchor).isActive = true
```
UI를 설정해주는 메서드를 만들어서 별도로 관리하는건 어떨까요?

hano-ios replied:
setUpView() 함수로 분리했습니다!!

wonhee009 reviewed (changes requested):
스텝2 수고하셨습니다!🎉
코멘트 남겨두었으니 확인 부탁드립니다!

### 화면

![image](https://user-images.githubusercontent.com/30456206/149123845-d2520cc3-1662-4bf1-aa71-67fb2bba00f6.png)

iPhone11 (iOS 15)에서 빌드 첫 화면에 텍스트가 보이지 않는데요. (+ 출품작 상세 화면도 동일합니다)
확인 부탁드립니다!

![image](https://user-images.githubusercontent.com/30456206/149123912-e2eb2800-5737-4545-9f24-0903c7b9216b.png)

"메인 화면 -> 출품작 리스트 -> 메인 화면 "으로 돌아올 경우 네비게이션 바가 위의 스크린샷처럼 노출되는데요.
확인 부탁드립니다!

![image](https://user-images.githubusercontent.com/30456206/149124447-ce0d8d5c-3834-4d0c-9355-02056bd8fb0d.png)
![image](https://user-images.githubusercontent.com/30456206/149124467-2eca95e5-fff0-4d72-8ab4-0c2769c9614a.png)

리스트 최상단에 위치해 있을때에는 네비바 배경색이 흰색인데 리스트를 살짝 내리면 배경색이 변경됩니다.
확인 부탁드립니다!

### 질문

- 화면과 관련한 질문은 detailView 쪽에 코멘트 남겨두었으니 확인 부탁드립니다!🙏

- [공식 문서](https://developer.apple.com/documentation/objectivec/nsobject/1402907-awakefromnib)에 따르면 
> Prepares the receiver for service after it has been loaded from an Interface Builder archive, or nib file.

Interface Builder나 nib 파일에 의해서 로드될 때에 호출된다고 하네요:)
(awakeFromNib와 init에 대해 비교 혹은 정리해둔 글은 많이 찾아보실 수 있을거에요!)

Commit:
refactor: 2차 PR 반영

함수, 변수, 상수 형 바꾸기,  리네이밍
스크롤뷰 크기 조정
필요없는 코드 삭제

Commit:
feat: initializer를 통한 데이터 초기화

DetailView와 DetailViewController의 init을 통해 데이터 초기화
-> 상수로 데이터 처리 가능

Commit:
refactor: 메인화면 뷰 수정

스크롤뷰 내부 크기 조정
exposition infomation 분리

Commit:
chore: 주석 삭제

Commit:
chore: [수정] 접근자, 함수 위치 수정

외부에서 접근하지 않아야 할 변수/함수는 private로 선언
init 함수는 클래스 최상단으로 이동했으며
tableViewDelegate, DataSource에 적절한 함수 위치를 찾아 이동했음

hano-ios commented:
@wonhee009 
2차 PR 코멘트 주신 사항 반영해서 수정했습니다!!
꼼꼼한 코멘트 정말 감사드립니다😮
또 수정해야할 사항 있으면 피드백 부탁드립니다!! 감사합니다!!!

wonhee009 commented on Expo1900/Expo1900/View/MainView.swift:
```diff
@@ -0,0 +1,165 @@
+        expositionDescription.text = expositionData.description
+    }
+    
+    private func settingStackView() {
```
여기도 다른 곳들과 통일성을 맞춰서 `setup-`은 어떨까요?

cheshire-kwon replied:
감사합니다 해당 함수를 놓치고 있었네요! 수정했습니다!

wonhee009 commented on Expo1900/Expo1900/View/MainView.swift:
```diff
@@ -0,0 +1,165 @@
+    private func settingStackView() {
+        self.addArrangedSubview(expositionTitle)
+        self.addArrangedSubview(expositionImageView)
+        self.addArrangedSubview(expositionVisitorStack)
+        self.addArrangedSubview(expositionLocationStack)
+        self.addArrangedSubview(expositionDurationStack)
```
'방문객: 000명" 이걸 구현하기 위해 stack에 Label을 두개 넣어서 해결하신 것 같습니다.
```swift
let text = "방문객: \(visitorsCount)명"
```
이렇게도 text를 조합해서 사용할 수 있는데요.
이런 방식이면 stackView를 줄일수 있고, 좀 더 간결하게 화면을 구성할 수 있을거 같습니다.

cheshire-kwon replied:
앞에있는 레이블과 데이터의 글씨 크기가 서로 달라 레이블 두개를 넣어서 구현했습니다!

wonhee009 replied:
[이런 것](https://developer.apple.com/documentation/foundation/attributedstring)도 확인해보시면 좋을거 같습니다!:)

hano-ios replied:
신기한 꿀팁 감사합니다! 수정했습니다!!

wonhee009 commented on Expo1900/Expo1900/ViewController/MainViewController.swift:
```diff
@@ -0,0 +1,49 @@
+        
+        mainView.topAnchor.constraint(equalTo: scrollView.topAnchor).isActive = true
+        mainView.bottomAnchor.constraint(equalTo: scrollView.bottomAnchor).isActive = true
+        mainView.widthAnchor.constraint(equalTo: scrollView.widthAnchor, constant: -50).isActive = true
+        
+        mainView.centerXAnchor.constraint(equalTo: scrollView.centerXAnchor).isActive = true
```
```swift
mainView.leadingAnchor.constraint(equalTo: scrollView.leadingAnchor, constant: 25).isActive = true
```
이렇게도 가능하겠네요:)

cheshire-kwon replied:
아.. 이런방법도 있군여! 감사합니다 ㅎㅎ

wonhee009 reviewed (changes requested):
몇몇 코멘트 남겨두었으니 확인부탁드립니다!🙏

![image](https://user-images.githubusercontent.com/30456206/149315858-fd753bb3-cb1f-468b-91b5-9994e4a81e74.png)

해금, 나전칠기의 경우 화면이 깨지는거 같은데요 확인 부탁드립니다!🙏
(iPhone11)

![image](https://user-images.githubusercontent.com/30456206/149316002-508ffbf8-f5dd-4197-8786-e1e81cb291ae.png)

list 클릭 후 detailView로 이동 후 되돌아 오면 클릭된 cell이 표시되는데요. 
이 부분이 어색한거 같습니다.
좀 더 자연스럽게 클릭한 후에는 didSelected를 풀어주는 방향은 어떨까요?

wonhee009 commented on Expo1900/Expo1900/ViewController/EntryListViewController.swift:
```diff
@@ -0,0 +1,45 @@
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: EntryTableViewCell.identifier) as? EntryTableViewCell ?? EntryTableViewCell()
```
`tableView.dequeueReusableCell(withIdentifier: EntryTableViewCell.identifier)`에서 원하는 cell이 제대로 반환되고 있나요?👀

cheshire-kwon replied:
identifier 등록을 안했었네요 ..ㅠㅠ 등록 추가하여 반환되는 셀 확인했습니다! 감사합니다.

Commit:
refactor: 버그 수정 및 리네이밍

디테일 화면 이미지 높이 추가
setting -> setup 함수이름 변경
특정 스택뷰 상수로 변경
cell identifer 등록

cheshire-kwon commented:
@wonhee009 
안녕하세요! 꼼꼼한 코드리뷰 정말 감사드립니다! 코멘트 주신 내용 반영하였으니 확인 부탁드립니다.
관련내용 댓글로 달아 두었고 추가로 테이블 셀의 셀렉트도 풀어주었습니다!

그리고 해금, 나전칠기 이미지를 확인해보니 다른 이미지들 보다 사이즈가 커서 해당 현상이 나타난거 아닌가 싶습니다.
우선은 heightanchor를 추가해두기 했는데, 이런 경우처럼 특정 이미지 크기가 나머지 이미지와 다를 경우 어떻게 처리해주는 것이 좋은 방법일까요?

감사합니다!! @HanEun-ji

wonhee009 commented on Expo1900/Expo1900/ViewController/MainViewController.swift:
```diff
@@ -20,7 +20,7 @@ class MainViewController: UIViewController {
         mainView.widthAnchor.constraint(equalTo: scrollView.widthAnchor, constant: -50).isActive = true
         
-        mainView.centerXAnchor.constraint(equalTo: scrollView.centerXAnchor).isActive = true
+        mainView.leadingAnchor.constraint(equalTo: scrollView.leadingAnchor, constant: 25).isActive = true
```
trailingAnchor까지 지정해준다면 widthAnchor 없이도 원하시는대로 동작할겁니다:)

hano-ios replied:
안녕하세요 이부분 고민이 많이 생겨서 한번만 더 질문드립니다 ㅠㅠ

```Swift
        mainView.topAnchor.constraint(equalTo: scrollView.topAnchor).isActive = true
        mainView.bottomAnchor.constraint(equalTo: scrollView.bottomAnchor).isActive = true
//        mainView.widthAnchor.constraint(equalTo: scrollView.frameLayoutGuide.widthAnchor, constant: -20).isActive = true
        mainView.leadingAnchor.constraint(equalTo: scrollView.frameLayoutGuide.leadingAnchor, constant: 10).isActive = true
        mainView.trailingAnchor.constraint(equalTo: scrollView.frameLayoutGuide.trailingAnchor, constant: -10).isActive = true
```
말씀대로 이렇게 수정해보았는데, width를 주지 않으면 레이어에 'Scrollable content size is ambiguous for UIScrollView.' 라는 에러가 떠서 width도 필요한게 아닌가 생각이 듭니다!! 
또, scrollView.frameLayout.trailingAnchor이 아니라 scrollView.trailingAnchor로 하게 되면 화면이 가로로 스크롤되는데 이부분도 왜그런지 이유를 잘 모르겠습니다..ㅠㅠ

wonhee009 replied:
<img width="1103" alt="image" src="https://user-images.githubusercontent.com/30456206/149467166-fe642dbc-657a-43ac-a938-b9613c1d62a1.png">
<img width="1103" alt="image" src="https://user-images.githubusercontent.com/30456206/149467206-e05f11d5-3522-4e76-a28d-c28a9c99cb28.png">

적어주신 코드 확인해봤는데요.
화면도 제대로 나오고 에러도 안 잡히는거 같아요🥲
혹시 좀 더 자세한 환경이나 설명 가능하실까요?.?

hano-ios replied:
![스크린샷 2022-01-14 오후 4 22 48](https://user-images.githubusercontent.com/24950485/149467744-a50fbaa5-f92a-4a9a-be62-bd51a09bb735.png)
실행 후 view hierarchy를 통해 확인했는데 이렇게 보라색 느낌표가 떠서 질문드렸습니다!
기능 자체나 화면에는 이상이 없지만 경고가 떠서 width 정보가 필요한건가 했습니당
(width 코드 적용했더니 느낌표가 사라지긴 했습니다!)

wonhee009 reviewed (approved):
@yeokyeong46 @HanEun-ji 스텝2 수고하셨습니다!🎉
남긴 코멘트들 잘 적용하셨네요👍

보통 이미지 같은 경우는 height 혹은 width 중 어떤 걸 기준으로 할지 가이드가 내려오는 편인거 같아요.
(혹은 둘 다 정해지는 경우도 있고요.)
현재 프로젝트에서 디자인 가이드로 따로 나와 있지 않다면 height, width 모두 지정해줘도 괜찮을거 같습니다!

몇몇 코멘트들도 다음 스텝에서 적용하기로 하고 이번 스텝은 여기서 마무리하겠습니다! 수고하셨어요~
